### ISSUE TITLE
FieldWrapper wraps the final element instead of the overall field type

### ISSUE THREAD

**Issue by richiejd** [dotansimha/graphql-code-generator#4614]
Basic example:
- expected: `FieldWrapper<Maybe<Array<Maybe<Favorite>>>>`
- actual: `Maybe<Array<Maybe<FieldWrapper<Address>>>>`

Not sure whether to call this a bug or just a missing feature, but the FieldWrapper is put inside rather than outside of the Array and Maybe wrappers.

Example:
With the config: 
```
      defaultMapper: Partial<{T}>
      fieldWrapperValue: T | Promise<T>
      wrapFieldDefinitions: true
```

Given the following graphql (note I'm contriving an example, but don't usually have nullables in an array...):

```
type User {
  addresses: [Address!]!
  orders: [Order]!
  saved: [Saved!]
  favorites: [Favorite]
  defaultAddress: Address!
  favoriteAddress: Address
}
```

I would expect the following ts:
```
export type User = {
  __typename?: "User";
  addresses: FieldWrapper<Array<Address>>;
  orders: FieldWrapper<Array<Maybe<Order>>>;
  saved?: FieldWrapper<Maybe<Array<Saved>>>;
  favorites?: FieldWrapper<Maybe<Array<Maybe<Favorite>>>>;
  defaultAddress: FieldWrapper<Address>;
  favoriteAddress?: FieldWrapper<Maybe<Address>>;
};
```

instead I get:
```
export type User = {
  __typename?: "User";
  addresses: Array<FieldWrapper<Address>>;
  orders: Array<Maybe<FieldWrapper<Address>>>;
  saved?: Maybe<Array<FieldWrapper<Address>>>;
  favorites?: Maybe<Array<Maybe<FieldWrapper<Address>>>>;
  defaultAddress: FieldWrapper<Address>;
  favoriteAddress?: Maybe<FieldWrapper<Address>>;
};
```

As you can see, I would expect support for FieldWrapper to wrap the whole thing so I can have `Promise<T> | T`

This is on version `1.17.0`

**Comment by richiejd** [comment]
This same issue is actually a further issue in the resolver map, e.g. I get the following errors
```
Type '(_: {}, __: {}, { customerId }: Context) => Promise<CustomerRecord | null>' is not assignable to type 'ResolverFn<Maybe<ResolverTypeWrapper<Partial<Customer>>>, {}, Context, {}>'.
  Type 'Promise<CustomerRecord | null>' is not assignable to type 'Partial<Customer> | Promise<Partial<Customer>> | Promise<Maybe<ResolverTypeWrapper<Partial<Customer>>>> | null'.
    Type 'Promise<CustomerRecord | null>' is not assignable to type 'Promise<Partial<Customer>>'.
      Type 'CustomerRecord | null' is not assignable to type 'Partial<Customer>'.
        Type 'null' is not assignable to type 'Partial<Customer>'.
```

I'm not sure exactly how it should be configured, but the `Maybe` being where it is causes a mismatch on the null type.
It should be `ResolverTypeWrapper<Maybe<Partial<Customer>>>` to fix the issue.

**Comment by hampusborgos** [comment]
Did you figure out a way to fix this? It makes the entire FieldWrapper pointless to use for more complicated models.

**Comment by hampusborgos** [comment]
Did you figure out a way to fix this? It makes the entire FieldWrapper pointless to use for more complicated models.

ie. having an `async` field on a model that returns an `Array<T | null>` is not really an uncommon scenario. Is there really no way to make this library output code that supports `async` fields? :(

@dotansimha Any input on how to use the generated code, when you have models that have Promise fields?

**Comment by hampusborgos** [comment]
I have attempted to solve this in a new PR: https://github.com/dotansimha/graphql-code-generator/pull/5753 that adds a new configuration option target to arrays (and `Maybe<>`) wrappers.

**Comment by dotansimha** [comment]
Fix is available in `@graphql-codegen/typescript@1.22.0`. Thanks @hampusborgos !
